### PR TITLE
 Catch Exception rather than Throwable to prevent an invalid write state

### PR DESCRIPTION
**Pull request by duro1**
Receiving an OutOfMemoryError in the HollowWriteStateEngine.resetToLastPrepareForNextCycle method resulted in having an empty write state. After this occurred, future snapshots published as if they were the initial snapshot, losing all previous states data.

Allow JVM level Errors to propagate rather than be wrapped into a RuntimeException that could more easily be incorrectly logged and ignored

### CODE REVIEW

**Comment by duro1** [comment]
@toolbear @akhaku Please take a look at the PR comments for the motivation of this change and let me know if you have any comments/concerns.
CC: @rpalcolea @tinetti

File: hollow/src/main/java/com/netflix/hollow/api/consumer/fs/HollowFilesystemAnnouncementWatcher.java [hunk 1]
```diff
@@ -132,8 +132,8 @@ public void run() {
                                 consumer.triggerAsyncRefresh();
                         }
                     }
-                } catch (Throwable th) {
-                    log.log(Level.WARNING, "Exception reading the current announced version", th);
+                } catch (Exception ex) {
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/api/consumer/fs/HollowFilesystemAnnouncementWatcher.java]: I wonder if we want this idiom:

```java
try {
  // ...
} catch (Exception ex) {
  log.log(WARNING, ...);
} catch (Throwable th) {
  log.log(ERROR, ...);
  throw th;
}
```

Just to leave a little clearer breadcrumb behind.

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 2]
```diff
@@ -192,8 +192,8 @@ public void run() {
             }
 
             executor.awaitSuccessfulCompletion();
-        } catch(Throwable t) {
-            throw new RuntimeException(t);
+        } catch(Exception ex) {
+            throw new RuntimeException(ex);
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java]: We've started replacing rethrowing as `HollowException` or a subclass of `HollowException` just to make forensics a little clearer. We change them as we come across them while making other related changes.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by duro1 — COMMENTED** [review]
@toolbear Thanks for the review comments. PR has been updated.  Please take another look

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 3]
```diff
@@ -166,7 +167,7 @@ public void run() {
         try {
             executor.awaitSuccessfulCompletion();
         } catch(Exception e){
-            throw new RuntimeException(e);
+            throw new HollowWriteStateException("Unable to restore write state from read state engine", e);
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:170]: 👍

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 4]
```diff
@@ -192,8 +193,8 @@ public void run() {
             }
 
             executor.awaitSuccessfulCompletion();
-        } catch(Throwable t) {
-            throw new RuntimeException(t);
+        } catch(Exception ex) {
+            throw new HollowWriteStateException("Failed to prepare for write", ex);
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:197]: 👍

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 5]
```diff
@@ -222,8 +223,8 @@ public void run() {
             }
 
             executor.awaitSuccessfulCompletion();
-        } catch(Throwable t) {
-            throw new RuntimeException(t);
+        } catch(Exception ex) {
+            throw new HollowWriteStateException("Failed to prepare for next cycle", ex);
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:227]: 👍

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 6]
```diff
@@ -261,8 +262,8 @@ public void run() {
 
         try {
             executor.awaitSuccessfulCompletion();
-        } catch(Throwable th) {
-            throw new RuntimeException(th);
+        } catch(Exception ex) {
+            throw new HollowWriteStateException("Unable to reset to the prior version of the write state", ex);
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:266]: 👍

**Review by ghost — APPROVED** [review]

**Pull request merged**